### PR TITLE
REGR: fix DataFrame reduction with EA columns and numeric_only=True

### LINKED ISSUE #33256: REGR: DataFrame.mean(numeric_only=True) raises AttributeError on v1.0.3
#### Code Sample, a copy-pastable example if possible

```python
>>> import numpy as np
>>>
>>> import pandas as pd
>>>
>>> pd.__version__
'1.0.3'
>>>
>>> df_wide = pd.DataFrame(np.random.randint(1000, size=(1000, 100))).astype("Int64").copy()
>>>
>>> df_wide.mean(numeric_only=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\simon\pandas\pandas\core\generic.py", line 11215, in stat_func
    f, name, axis=axis, skipna=skipna, numeric_only=numeric_only
  File "C:\Users\simon\pandas\pandas\core\frame.py", line 7896, in _reduce
    res = df._data.reduce(op, axis=1, skipna=skipna, **kwds)
  File "C:\Users\simon\pandas\pandas\core\internals\managers.py", line 351, in reduce
    bres = func(blk.values, *args, **kwargs)
  File "C:\Users\simon\pandas\pandas\core\nanops.py", line 69, in _f
    return f(*args, **kwargs)
  File "C:\Users\simon\pandas\pandas\core\nanops.py", line 102, in f
    if values.size == 0 and kwds.get("min_count") is None:
AttributeError: 'IntegerArray' object has no attribute 'size'
>>>

```
#### Problem description

This is a regression from 0.25.3

0aa48f7d9269206dea492ed14d5dfc6f46468de3 is the first bad commit
commit 0aa48f7d9269206dea492ed14d5dfc6f46468de3
Author: jbrockmendel <jbrockmendel@gmail.com>
Date:   Wed Jan 1 09:18:20 2020 -0800

    PERF: perform reductions block-wise (#29847)

on master raises `AttributeError: 'int' object has no attribute 'dtype'`

```
>>> df_wide.mean(numeric_only=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\simon\pandas\pandas\core\generic.py", line 11114, in stat_func
    func, name=name, axis=axis, skipna=skipna, numeric_only=numeric_only
  File "C:\Users\simon\pandas\pandas\core\frame.py", line 7990, in _reduce
    res = df._data.reduce(blk_func)
  File "C:\Users\simon\pandas\pandas\core\internals\managers.py", line 362, in reduce
    bres = func(blk.values, *args, **kwargs)
  File "C:\Users\simon\pandas\pandas\core\frame.py", line 7985, in blk_func
    return op(values, axis=0, skipna=skipna, **kwds)
  File "C:\Users\simon\pandas\pandas\core\nanops.py", line 120, in f
    result = bn_func(values, axis=axis, **kwds)
  File "<__array_function__ internals>", line 6, in nanmean
  File "C:\Users\simon\Anaconda3\envs\pandas-dev\lib\site-packages\numpy\lib\nanfunctions.py", line 952, in nanmean
    avg = _divide_by_count(tot, cnt, out=out)
  File "C:\Users\simon\Anaconda3\envs\pandas-dev\lib\site-packages\numpy\lib\nanfunctions.py", line 219, in _divide_by_count
    return a.dtype.type(a / b)
AttributeError: 'int' object has no attribute 'dtype'
```

#### Expected Output

```
>>> import numpy as np
>>>
>>> import pandas as pd
>>>
>>> pd.__version__
'0.25.3'
>>>
>>> df_wide = pd.DataFrame(np.random.randint(1000, size=(1000, 100))).astype("Int64").copy()
>>>
>>> df_wide.mean(numeric_only=True)
0     520.057
1     507.735
2     501.618
3     506.590
4     501.500
       ...
95    507.594
96    483.273
97    506.330
98    497.068
99    508.118
Length: 100, dtype: float64
>>>
```

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes #33256

For the block-wise path in `DataFrame._reduce`, we ensure to use the EA reduction for blocks holding EAs, instead of passing the EA to the `op` function (which is typically the `nanops` nan function).

This needs https://github.com/pandas-dev/pandas/pull/33758 to be merged, since that PR fixes the test that is failing here.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -8281,10 +8281,10 @@ def _get_data(axis_matters):
             out_dtype = "bool" if filter_type == "bool" else None
 
             def blk_func(values):
-                if values.ndim == 1 and not isinstance(values, np.ndarray):
-                    # we can't pass axis=1
-                    return op(values, axis=0, skipna=skipna, **kwds)
-                return op(values, axis=1, skipna=skipna, **kwds)
+                if isinstance(values, ExtensionArray):
+                    return values._reduce(name, skipna=skipna, **kwds)
```
Comment (jbrockmendel) [inline pandas/core/frame.py:8285]: is _reduce part of the interface?

if we're avoiding nanops, can/should we make those functions either a) not handle EA cases or b) dispatch to _reduce?

**Comment by jorisvandenbossche** [comment]
> is _reduce part of the interface?

yes (and the default is to raise an error saying the reduction is not supported)

> if we're avoiding nanops, can/should we make those functions either a) not handle EA cases or b) dispatch to _reduce?

Since `_reduce` is the interface that Series calls, and thus IMO also DataFrame should call for EA columsn (which I am doing here for the block wise path, and trying to do in general in #32867 (also for `numeric_only=None`), then in principle, I think we can/should always call that for EAs, and should not need to call a nanops function for an EA. Which means that in principle nanops would not need to handle EAs.

Of course, we could also decide on always calling nanops, and putting the logic to call `_reduce` for EAs in nanops. That's a choice we can make. 
(but in either way, I think the core algorithmic code of nanops should not need to handle EAs, as the `EA._reduce` can pass the correct ndarray to nanops and reinterpret the result)

**Comment by jorisvandenbossche** [comment]
More comments on this? (apart from that I need to rebase + fix the tests (update error message match))

**Comment by jreback** [comment]
seems ok to me

**Comment by jbrockmendel** [comment]
ill take another look once rebased+green, but tentatively looks good

**Comment by jorisvandenbossche** [comment]
Updated on master

**Comment by jorisvandenbossche** [comment]
(travis failure is an unrelated URLError connection error in test_read_from_http_url)

**Comment by jbrockmendel** [comment]
LGTM

**Comment by simonjayhawkins** [comment]
xref #33300 to track.

**Pull request merged**